### PR TITLE
fix(validate-inputs): accept new stateless ghs_ installation token format

### PR DESCRIPTION
**Pull request by ivuorinen**
## Summary

- GitHub began rolling out a stateless installation-token format (`ghs_APPID_JWT`,
  ~520 chars with two dots and underscores) on 2026-04-27. The validator's old
  regex `^ghs_[a-zA-Z0-9]{36}$` rejected this, causing every action that calls
  `validate-inputs` with the runtime `github.token` to fail with
  "Invalid token format".
- Updates the regex in all four parallel implementations (production validator,
  secrets-leak detector, and the two test-framework mirrors) to GitHub's
  recommended pattern, and adds 22 new test cases covering both stateful and
  stateless formats with boundary checks.
- Cleans up three pre-existing bugs in `generate-tests.py` discovered while
  updating the template (wrong char counts, missing `required=True`, reference
  to non-existent `validate_pypi_token`), and removes an unused `config`
  parameter that was silenced with `# noqa: ARG002`.

## Affected actions

Every action that calls `validate-inputs` with the runtime GitHub token was
broken: `ansible-lint-fix`, `codeql-analysis`, `csharp-publish`, `pr-lint`,
`pre-commit`, `python-lint-fix`, `security-scan`, `stale`, `terraform-lint-fix`.

## Reference

- Official guidance:
  <https://github.blog/changelog/2026-05-15-github-app-installation-tokens-per-request-override-header>
- GitHub's recommended regex: `ghs_[A-Za-z0-9\._]{36,}`
- Note from <https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/about-authentication-to-github#githubs-token-formats>:
  > If your application expects or relies on installation tokens being exactly
  > 40 characters long, it may not handle this new token format correctly.

## Commits

1. `fix(validate-inputs): accept new stateless ghs_ installation token format`
   — core regex update across 4 parallel sites.
2. `test(validate-inputs): cover stateless ghs_ format with happy/sad cases`
   — adds 45 lines of test coverage to `test_token.py`, expands
   `version_test_data.py` fixtures with realistic JWT + boundary cases, adds
   two new test files in `_tests/` for the framework validators, fixes
   pre-existing generator template bugs.
3. `docs(audit): record N-096..N-099 ghs_ stateless format fixes`
   — appends Pass 14 and Pass 15 findings to `docs/audit/nitpicker-findings.md`.

## Test plan

- [x] Pytest: validate-inputs suite passes (786 tests).
- [x] Pytest: new test framework files pass (22 tests).
- [x] Boundary tests: body of exactly 36 and 1024 chars accepted; 35 and 1025
      rejected.
- [x] Char-class tests: `+`, `/`, ` ` all correctly rejected.
- [x] End-to-end: synthetic 460-char stateless JWT token validates cleanly.
- [x] Secrets-leak detector flags new format tokens embedded in unrelated
      inputs.
- [x] Ruff lint: clean.
- [ ] Optional: trigger one of the affected actions in CI with the runtime
      `github.token` to confirm the rollout token format is accepted.

## Release Notes

* **New Features**
  * Added support for GitHub App installation tokens in both stateful and stateless JWT-style formats.

* **Improvements**
  * Expanded GitHub token validation to accept a broader range of installation token formats and updated validation messaging.

* **Tests**
  * Added comprehensive test coverage for new token validation scenarios.

* **Documentation**
  * Updated audit documentation with latest validation findings.

[Review Change Stack](https://app.coderabbit.ai/change-stack/ivuorinen/actions/pull/587?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

File: validate-inputs/validators/token.py [hunk 1]
```diff
@@ -108,7 +114,8 @@ def validate_github_token(self, token: str, *, required: bool = False) -> bool:
         self.add_error(
             "Invalid token format. Expected: ghp_* (40 chars), "
             "github_pat_[A-Za-z0-9_]* (50-255 chars), gho_* (40 chars), ghu_* (40 chars), "
-            "ghs_* (40 chars), ghr_* (40 chars), ghe_* (40 chars), or ${{ github.token }}",
+            "ghs_* (40 chars stateful or ~520 chars stateless JWT), "
+            "ghr_* (40 chars), ghe_* (40 chars), or ${{ github.token }}",
```
[the remark this answers is not in the document]
Comment (ivuorinen) [reply validate-inputs/validators/token.py]: Fixed in 5808e46. The error message now explicitly lists `${{ ... }}` expressions (with `${{ github.token }}` and `${{ secrets.GITHUB_TOKEN }}` as examples) and `$VAR` / `${VAR}` env-var references as accepted non-literal forms, alongside the explicit regex pattern for each token type. The same defect existed in `_tests/framework/validation.py` and `_tests/shared/validation_core.py` — both fixed in the same commit per the codebase scan.

File: _tests/shared/validation_core.py [hunk 2]
```diff
@@ -113,7 +115,8 @@ def validate_github_token(self, token: str, *, required: bool = False) -> tuple[
         return (
             False,
             "Invalid token format. Expected: gh[efpousr]_* (36 chars), "
-            "github_pat_[A-Za-z0-9_]* (50-255 chars), ghs_* (36 chars), or npm_* (40+ chars)",
+            "github_pat_[A-Za-z0-9_]* (50-255 chars), "
+            "ghs_* (40 chars stateful or ~520 chars stateless JWT), or npm_* (40+ chars)",
         )
```
[the remark this answers is not in the document]
Comment (ivuorinen) [reply _tests/shared/validation_core.py:126]: Fixed in 5808e46. The message now reads `gh[efpousr]_[a-zA-Z0-9]{36} (40 chars total — ghp/gho/ghu/ghs/ghr/ghe)` so the regex is explicit and the total length is unambiguous. The accepted `${{ ... }}` expressions and `$VAR` env-var references are now listed too. Same fix applied to the structurally identical message in `_tests/framework/validation.py`.

File: _tests/framework/validation.py [hunk 3]
```diff
@@ -100,7 +102,8 @@ def validate_github_token(self, token: str, action_dir: str = "") -> tuple[bool,
         return (
             False,
             "Invalid token format. Expected: gh[efpousr]_* (36 chars), "
-            "github_pat_[A-Za-z0-9_]* (50-255 chars), ghs_* (36 chars), or npm_* (40+ chars)",
+            "github_pat_[A-Za-z0-9_]* (50-255 chars), "
+            "ghs_* (40 chars stateful or ~520 chars stateless JWT), or npm_* (40+ chars)",
         )
```
[the remark this answers is not in the document]
Comment (ivuorinen) [reply _tests/framework/validation.py:113]: Fixed in 5808e46. Structurally identical to the `_tests/shared/validation_core.py` thread — same rewrite applied: explicit regex `gh[efpousr]_[a-zA-Z0-9]{36} (40 chars total)`, explicit `ghs_[A-Za-z0-9._]{36,1024}` bounds for installation tokens, and an enumeration of all accepted `${{ ... }}` and `$VAR` non-literal forms.

File: _tests/framework/validation.py [hunk 4]
```diff
@@ -100,7 +102,8 @@ def validate_github_token(self, token: str, action_dir: str = "") -> tuple[bool,
         return (
             False,
             "Invalid token format. Expected: gh[efpousr]_* (36 chars), "
-            "github_pat_[A-Za-z0-9_]* (50-255 chars), ghs_* (36 chars), or npm_* (40+ chars)",
+            "github_pat_[A-Za-z0-9_]* (50-255 chars), "
+            "ghs_* (40 chars stateful or ~520 chars stateless JWT), or npm_* (40+ chars)",
```
[the remark this answers is not in the document]
Comment (ivuorinen) [reply _tests/framework/validation.py]: Fixed in 5808e46. The `ghs_` fragment now reads `ghs_[A-Za-z0-9._]{36,1024} (40-1028 chars total — installation; stateful or stateless JWT)`, matching the enforced regex bounds exactly. The full message was also restructured to use explicit regex syntax for every accepted form and to enumerate `${{ ... }}` / `$VAR` non-literal forms. Same fix mirrored in `_tests/shared/validation_core.py` and `validate-inputs/validators/token.py`.
